management/sql/client.go

```diff
@@ -0,0 +1,163 @@
+package sql
+
+import (
+	"encoding/xml"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/management"
+)
+
+// Definitions of numerous constants representing API endpoints.
+const (
+	azureCreateDatabaseServerURL = "services/sqlservers/servers"
+	azureListDatabaseServersURL  = "/services/sqlservers/servers"
+	azureDeleteDatabaseServerURL = "/services/sqlservers/servers/%s"
+
+	azureCreateDatabaseURL = "services/sqlservers/servers/%s/databases"
+	azureGetDatabaseURL    = "/services/sqlservers/servers/%s/databases/%s"
+	azureListDatabasesURL  = "/services/sqlservers/servers/%s/databases?contentview=generic"
+	azureDeleteDatabaseURL = "/services/sqlservers/servers/%s/databases/%s"
+
+	errParamNotSpecified = "Parameter %s was not specified."
+)
+
+// SqlDatabaseClient defines various database CRUD operations.
+// It contains a management.Client for making the actual http calls.
+type SqlDatabaseClient struct {
+	mgmtClient management.Client
+}
+
+// NewClient returns a new SqlDatabaseClient struct with the provided
+// management.Client as the underlying client.
+func NewClient(mgmtClient management.Client) SqlDatabaseClient {
+	return SqlDatabaseClient{mgmtClient}
+}
+
+// CreateServer creates a new Azure SQL Database server and return its name.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505699.aspx
+func (c SqlDatabaseClient) CreateServer(params DatabaseServerCreateParams) (string, error) {
+	req, err := xml.Marshal(params)
+	if err != nil {
+		return "", err
+	}
+
+	_, resp, err := c.mgmtClient.SendAzurePostRequestWithReturnedResponse(azureCreateDatabaseServerURL, req)
+	if err != nil {
+		return "", err
+	}
+
+	var name string
+	err = xml.Unmarshal(resp, &name)
+
+	return name, err
+}
+
+// ListServers retrieves the Azure SQL Database servers for this subscription.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505702.aspx
+func (c SqlDatabaseClient) ListServers() (ListDatabaseServersResponse, error) {
+	var resp ListDatabaseServersResponse
+
+	data, err := c.mgmtClient.SendAzureGetRequest(azureListDatabaseServersURL)
+	if err != nil {
+		return resp, err
+	}
+
+	err = xml.Unmarshal(data, &resp)
+	return resp, err
+}
+
+// DeleteServer deletes an Azure SQL Database server (including all its databases).
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505695.aspx
+func (c SqlDatabaseClient) DeleteServer(name string) error {
+	if name == "" {
+		return fmt.Errorf(errParamNotSpecified, "name")
+	}
+
+	url := fmt.Sprintf(azureDeleteDatabaseServerURL, name)
+	_, err := c.mgmtClient.SendAzureDeleteRequest(url)
+	return err
+}
+
+// CreateDatabase creates a new Microsoft Azure SQL Database on the given database server.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505701.aspx
+func (c SqlDatabaseClient) CreateDatabase(params DatabaseCreateParams, server string) error {
+	if params.Name == "" {
```

Comment:
do not check for `params.name`, it's not required to make the request, it's rather required to do the operation successfully and that's what API server checks in the input as well. so leave the check to the API server when you can.